### PR TITLE
feat(router): add route to invalidate cache entry

### LINKED ISSUE #995: [FEATURE] add a route that will invalidate cache
### Feature Description

Currently we don't have any way to manually invalidate cache that's stored in redis and in-memory, Add a route that given a key will invalidate both of redis and inmemory cache.

### Possible Implementation

So the in memory cache currently has `invalidate` function that will delete the entry in the cache, and in our redis interface we have `delete_key` function that will delete the entry from redis. You need to create a route in `app.rs` that will call this on `AdminApiKeyAuth`. 

### Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

None

### PR DESCRIPTION
**Pull request by jeevaprakashdr**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
A new route to invalidate the in-memory cache and redis

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Currently we don't have any way to manually invalidate cache that's stored in redis and in-memory, 

Closes #995.

## How did you test it?
A unit test is added to test the in-memory cache and postman script used for redis cache.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [ ] I addressed lints thrown by `cargo clippy`
- [ ] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/core/errors/api_error_response.rs [hunk 1]
```diff
@@ -182,6 +182,8 @@ pub enum ApiErrorResponse {
     MissingFilePurpose,
     #[error(error_type = ErrorType::InvalidRequestError, code = "HE_04", message = "File content type not found / valid")]
     MissingFileContentType,
+    #[error(error_type = ErrorType::RouterError, code = "RE_01", message = "Failed to remove redis cache key")]
+    RemoveRedisKeyFailure,
```
Comment (SanchithHegde) [inline crates/router/src/core/errors/api_error_response.rs:186]: @dracarys18 @Narayanbhat166 Please confirm if we would need a new error to be introduced for this scenario, or if we would prefer an `InternalServerError` being raised instead.

File: .gitignore [hunk 2]
```diff
@@ -258,3 +258,5 @@ loadtest/*.tmp/
 result*
 
 .idea/
+
+.vscode/settings.json
```
Comment (SanchithHegde) [inline .gitignore:262]: Could you please revert this change? We have ignored most files under the `.vscode` directory. This specific file could be used for sharing project specific settings across developers working on the project, on a later date.

In fact, if you think such a thing would be helpful for you, I'll include project specific settings that can be followed across by developers working on the project, in a separate PR.

File: crates/router/src/core/cache.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+use common_utils::errors::CustomResult;
+
+use super::errors;
+use crate::{
+    cache::CONFIG_CACHE,
+    db::StorageInterface,
+    services::{self},
```
Comment (SanchithHegde) [inline crates/router/src/core/cache.rs:7]:
```suggestion
    services,
```

File: crates/router/src/core/cache.rs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+pub async fn invalidate(
+    store: &dyn StorageInterface,
+    key: &str,
+) -> CustomResult<services::api::ApplicationResponse<serde_json::Value>, errors::ApiErrorResponse> {
+    CONFIG_CACHE.remove(key).await;
+
+    match store.get_redis_conn().delete_key(key).await {
+        Ok(_) => Ok(services::api::ApplicationResponse::StatusOk),
+        _ => Err(errors::ApiErrorResponse::RemoveRedisKeyFailure.into()),
```
Comment (SanchithHegde) [inline crates/router/src/core/cache.rs:18]: Preferably, we should be returning a response with 404 status code in case a key was not found, and a response with 500 status code if a Redis related issue occurred.

File: crates/router/src/routes/cache.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+#[instrument(skip_all)]
+pub async fn invalidate(
+    state: web::Data<AppState>,
+    req: HttpRequest,
+    key: web::Path<String>,
+) -> impl Responder {
+    let flow = Flow::CacheInvalidate;
+
+    let key = key.into_inner().to_owned();
+    println!("cache key is {}", &key);
```
Comment (SanchithHegde) [inline crates/router/src/routes/cache.rs:19]: Please remove this `println!()` statement.

File: crates/router/tests/cache.rs
Comment (SanchithHegde) [inline crates/router/tests/cache.rs:1]: Once you've updated the code to return a response with a 400 status code for an invalid key, please add a test case for it.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/core/errors/api_error_response.rs (hunk 1 above)
Comment (Narayanbhat166) [reply crates/router/src/core/errors/api_error_response.rs:186]: we can use InternalServerError in its place. This function/route does only one thing, that is to remove the key from redis, there are not a lot of moving parts. This error tells that `RedisKeyRemovalFailed`, in order to know why it failed we anyway have to look into the logs and debug. So This error will not add much value over using `InternalServerError`.

File: crates/router/src/core/cache.rs (hunk 3 above)
Comment (jeevaprakashdr) [reply crates/router/src/core/cache.rs:7]: good spot!

File: crates/router/src/core/cache.rs (hunk 4 above)
Comment (jeevaprakashdr) [reply crates/router/src/core/cache.rs:18]: I am happy to return the InternalServerError(HTTP 500) as part of the error. 

In case of the key does not exist what needs to be done may be up for discussion. As of now RedisConnectionPool's `del(key)` encapsulation states  

> Removes the specified keys. A key is ignored if it does not exist. --> Based on this currently we are returning HTTP 200 status code. 

If we need to return the HTTP 404 status code,  need to extend `RedisConnectionPool::delete_key`.  This has currently been referenced in a couple of places.

File: crates/router/tests/cache.rs
Comment (jeevaprakashdr) [reply crates/router/tests/cache.rs:1]: can be done based on the outcome of 
[previous sugestion ](https://github.com/juspay/hyperswitch/pull/1100#discussion_r1190380458)

File: .gitignore (hunk 2 above)
Comment (jeevaprakashdr) [reply .gitignore]: of course, I can. If i do come across this often, will remove a new issue.

File: crates/router/src/core/cache.rs (hunk 4 above)
Comment (SanchithHegde) [reply crates/router/src/core/cache.rs]: @dracarys18 @Narayanbhat166 What should be the behavior here?

Comment (Narayanbhat166) [reply crates/router/src/core/cache.rs]: We can change this function https://github.com/juspay/hyperswitch/blob/cef8914372fa051f074e89fc76b76c6aee0d7bca/crates/redis_interface/src/commands.rs#L151 
to return a https://github.com/juspay/hyperswitch/blob/cef8914372fa051f074e89fc76b76c6aee0d7bca/crates/redis_interface/src/errors.rs#L52 error, and then the caller can match based on the error type?

Comment (dracarys18) [reply crates/router/src/core/cache.rs]: @SanchithHegde @jeevaprakashdr As I can see in the fred documentation del will ignore if key does not exist in the database so it won't throw an error, So if del failed possibility is that something went wrong with redis so makes sense to throw InternalServerError.

Ref: https://docs.rs/fred/latest/fred/interfaces/trait.KeysInterface.html#method.del

Comment (SanchithHegde) [reply crates/router/src/core/cache.rs]: @dracarys18 That seems to be how Redis handles it as well (official docs for the [`DEL` command](https://redis.io/commands/del/)). However, since we are only deleting one key, the output of the `DEL` command as an integer must be either 0 or 1. If we absolutely need to return a 404, we can match over the response (0 or 1), and return an appropriate result (`Ok(())` or `Err(NotFound)`). If we get any other response than 0 or 1, we'd have to error out stating an unexpected response from Redis.

Comment (Narayanbhat166) [reply crates/router/src/core/cache.rs]: > @dracarys18 That seems to be how Redis handles it as well (official docs for the [`DEL` command](https://redis.io/commands/del/)). However, since we are only deleting one key, the output of the `DEL` command as an integer must be either 0 or 1. If we absolutely need to return a 404, we can match over the response (0 or 1), and return an appropriate result (`Ok(())` or `Err(NotFound)`). If we get any other response than 0 or 1, we'd have to error out stating an unexpected response from Redis.

This seems to be a good option.

Comment (Narayanbhat166) [reply crates/router/src/core/cache.rs]: > @SanchithHegde @jeevaprakashdr As I can see in the fred documentation del will ignore if key does not exist in the database so it won't throw an error, So if del failed possibility is that something went wrong with redis so makes sense to throw InternalServerError.
> 
> Ref: https://docs.rs/fred/latest/fred/interfaces/trait.KeysInterface.html#method.del

Returning 200 will be ambiguous in this case, you are not sure whether the key was deleted or the key was given in the wrong format ( key not found ). It would be better if the end user is known that key was not found and hence the deletion did not happen

File: crates/redis_interface/src/commands.rs [hunk 6]
```diff
@@ -147,13 +147,17 @@ impl super::RedisConnectionPool {
...
-        self.pool
+        match self.pool
             .del(key)
-            .await
-            .into_report()
-            .change_context(errors::RedisError::DeleteFailed)
+            .await{
+                Ok(0) | Ok(i32::MIN..=-1_i32) => Err(errors::RedisError::NotFound).into_report(),
+                Ok(1_i32..=i32::MAX) =>  Ok(()),
+                Err(fred::error::RedisError {..}) => return Err(errors::RedisError::DeleteFailed).into_report(),
```
Comment (SanchithHegde) [inline crates/redis_interface/src/commands.rs:159]: We'd prefer matching against 0 and 1 only, while keeping the other values as unexpected responses. In the previous conversation, I forgot that we were doing similar response handling elsewhere in the `redis_interface` crate, in a slightly neater way. Could you define an enum `DelReply` like so, which implements the `fred::types::FromRedis` trait:

```rust
enum DelReply {
    KeyDeleted,
    KeyNotDeleted, // Key not found
}
```

You can refer to this snippet for implementing the `FromRedis` trait:

https://github.com/juspay/hyperswitch/blob/d7cfb4a179083580a7e195fa07077af23a262ceb/crates/redis_interface/src/types.rs#L135-L152

Then, you can match on the `DelReply` enum and return either an `Ok(())`, `Err(RedisError::NotFound)`, or `Err(RedisError::DeleteFailed)`.

File: crates/router/src/core/cache.rs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+pub async fn invalidate(
+    store: &dyn StorageInterface,
+    key: &str,
+) -> CustomResult<services::api::ApplicationResponse<serde_json::Value>, errors::ApiErrorResponse> {
+    CONFIG_CACHE.remove(key).await;
+
+    match store.get_redis_conn().delete_key(key).await {
+        Ok(_) => Ok(services::api::ApplicationResponse::StatusOk),
+        _ => Err(errors::ApiErrorResponse::InternalServerError.into()),
+    }
```
Comment (SanchithHegde) [inline crates/router/src/core/cache.rs:15]: This would have to return a response with a 404 status code when the key is not deleted.

File: crates/router/src/cache.rs [hunk 8]
```diff
@@ -63,6 +63,10 @@ impl Cache {
         let val = self.get(key)?;
         (*val).as_any().downcast_ref::<T>().cloned()
     }
+
+    pub async fn remove(&self, key: &str) {
```
Comment (dracarys18) [inline crates/router/src/cache.rs:67]: Hey @jeevaprakashdr can you also replace it `invalidate` with `remove` in the other places where `invalidate` is being used

Comment (jeevaprakashdr) [reply crates/router/src/cache.rs:67]: @dracarys18  ya can do. then in that case following will be the changes. is this okay?
*Url Endpoint will be renamed 
               **From**  http://localhost:8080/cache/invalidate/{key}
               **To** http://localhost:8080/cache/remove/{key}
*core/cache.rs method is renamed 
            **from** invalidate(key)
            **to** remove(key)

File: crates/redis_interface/src/commands.rs [hunk 9]
```diff
@@ -21,7 +21,7 @@ use fred::{
         RedisKey, RedisMap, RedisValue, Scanner, SetOptions, XCap, XReadResponse,
     },
 };
-use futures::StreamExt;
+use futures::{StreamExt};
```
[the remark this answers is not in the document]
Comment (jeevaprakashdr) [reply crates/redis_interface/src/commands.rs]: done

File: crates/redis_interface/src/commands.rs (hunk 6 above)
Comment (jeevaprakashdr) [reply crates/redis_interface/src/commands.rs]: done

File: crates/router/src/cache.rs (hunk 8 above)
Comment (dracarys18) [reply crates/router/src/cache.rs:125]: It's fine to keep it as invalidate @jeevaprakashdr

**Comment by dracarys18** [comment]
Can you add invalidate for `ACCOUNTS_CACHE` as well since [this](https://github.com/juspay/hyperswitch/commit/da4d721424d329af618a63034aabe2d9248eb041) change.

**Comment by jeevaprakashdr** [comment]
@Narayanbhat166 @jarnura @ashokkjag  The changes are ready for review. Please can you have a look at the PR, when you get a chance.

File: crates/router/src/core/cache.rs [hunk 10]
```diff
@@ -0,0 +1,16 @@
+use common_utils::errors::CustomResult;
+
+use super::errors;
+use crate::{cache::CONFIG_CACHE, db::StorageInterface, services};
+
+pub async fn invalidate(
+    store: &dyn StorageInterface,
+    key: &str,
+) -> CustomResult<services::api::ApplicationResponse<serde_json::Value>, errors::ApiErrorResponse> {
+    CONFIG_CACHE.remove(key).await;
```
Comment (dracarys18) [inline crates/router/src/core/cache.rs:10]: @jeevaprakashdr can you add `ACCOUNT_CACHE.remove(key).await`

Comment (jeevaprakashdr) [reply crates/router/src/core/cache.rs:14]: ahh okie, now it is clear a new in-memory cache has been introduced as part of  [this change](https://github.com/juspay/hyperswitch/commit/da4d721424d329af618a63034aabe2d9248eb041) and it should be invalidated through the new endpoint. 

The changes are done and pushed.

File: crates/drainer/src/utils.rs [hunk 11]
```diff
@@ -100,6 +100,7 @@ pub async fn make_stream_available(
         .delete_key(stream_name_flag)
         .await
         .map_err(DrainerError::from)
+        .map(|_r| ())
```
Comment (SanchithHegde) [inline crates/drainer/src/utils.rs:103]: Please revert this change.

File: crates/redis_interface/src/commands.rs [hunk 12]
```diff
@@ -148,12 +148,17 @@ impl super::RedisConnectionPool {
...
+        let result: Result<DelReply, fred::error::RedisError> = self.pool.del(key).await;
+        match result {
+            Ok(del_reply) => Ok(del_reply),
+            Err(e) => match e.kind() {
+                fred::error::RedisErrorKind::NotFound => {
+                    Err(errors::RedisError::NotFound).into_report()
+                }
+                _ => Err(errors::RedisError::DeleteFailed).into_report(),
+            },
+        }
```
Comment (SanchithHegde) [inline crates/redis_interface/src/commands.rs:161]:
```suggestion
        self.pool.del(key).await.into_report().change_context(errors::RedisError::DeleteFailed)
```

You should be able to do something similar, since the `NotFound` error is never raised in the current implementation.

File: crates/redis_interface/src/commands.rs [hunk 13]
```diff
@@ -665,29 +670,81 @@ impl super::RedisConnectionPool {
 #[cfg(test)]
 mod tests {
     #![allow(clippy::unwrap_used)]
+    #![allow(clippy::expect_used)]
```
Comment (SanchithHegde) [inline crates/redis_interface/src/commands.rs:673]:
```suggestion
    #![allow(clippy::expect_used, clippy::unwrap_used)]
```

File: crates/router/src/core/cache.rs [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+    ACCOUNTS_CACHE.remove(key).await;
+
+    match store.get_redis_conn().delete_key(key).await {
+        Ok(_del_reply) => Ok(services::api::ApplicationResponse::StatusOk),
+        _ => Err(errors::ApiErrorResponse::InvalidRequestUrl.into()),
```
Comment (SanchithHegde) [inline crates/router/src/core/cache.rs:19]: Here, we can match on `del_reply` and return a 200, 404 or 500 status code, something like this:

```rust
Ok(DelReply::KeyDeleted) => Ok(services::api::ApplicationResponse::StatusOk),
Ok(DelReply::KeyNotDeleted) => Err(errors::ApiErrorResponse::InvalidRequestUrl.into()),
error => error.into_report().change_context(errors::ApiErrorResponse::InternalServerError).attach_printable("Failed to invalidate cache")
```

You'd have to run the formatter by the way.

File: crates/router/tests/cache.rs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        ))
+        .insert_header(api_key)
+        .send()
+        .await
+        .unwrap();
+
+    // Assert
+    let response_body = response.body().await;
+    println!("invalidate Cache: {response:?} : {response_body:?}");
+    assert_eq!(response.status(), awc::http::StatusCode::OK);
```
Comment (SanchithHegde) [inline crates/router/tests/cache.rs:75]: Once you update the logic for the invalidation endpoint, this would then become `StatusCode::NotFound`.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/core/cache.rs (hunk 14 above)
Comment (jeevaprakashdr) [reply crates/router/src/core/cache.rs]: done

File: crates/router/tests/cache.rs (hunk 15 above)
Comment (jeevaprakashdr) [reply crates/router/tests/cache.rs]: good spot

File: crates/drainer/src/utils.rs (hunk 11 above)
Comment (SanchithHegde) [reply crates/drainer/src/utils.rs:103]: @jeevaprakashdr This is still pending.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/drainer/src/utils.rs (hunk 11 above)
Comment (jeevaprakashdr) [reply crates/drainer/src/utils.rs:103]: I thought I have dropped a message,  on this.  WE would need the error part  else we get the following error 

image

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/cache.rs [hunk 16]
```diff
@@ -131,17 +135,25 @@ mod cache_tests {
...
     }
 
     #[tokio::test]
-    async fn eviction_on_size_test() {
-        let cache = Cache::new(2, 2, Some(0));
+    async fn invalidate_cache_for_key() {
+        let cache = Cache::new(1800, 1800, None);
         cache.push("key".to_string(), "val".to_string()).await;
+
+        cache.remove("key").await;
```
Comment (dracarys18) [inline crates/router/src/cache.rs:149]: Can you add the assert it with for value present since this is a test.

**Review by dracarys18 — COMMENTED** [review]
Other than LGTM

File: crates/router/src/cache.rs (hunk 16 above)
Comment (jeevaprakashdr) [reply crates/router/src/cache.rs:149]: the changes are pushed now.

**Review by dracarys18 — APPROVED** [review]

**Review by SanchithHegde — DISMISSED** [review]

File: crates/router/src/core/cache.rs [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    CONFIG_CACHE.remove(key).await;
+    ACCOUNTS_CACHE.remove(key).await;
+
+    match store.get_redis_conn().delete_key(key).await {
+        Ok(DelReply::KeyDeleted) => Ok(services::api::ApplicationResponse::StatusOk),
+        Ok(DelReply::KeyNotDeleted) => Err(errors::ApiErrorResponse::InvalidRequestUrl.into()),
+        Err(error) => Err(error
+            .change_context(errors::ApiErrorResponse::InternalServerError)
+            .attach_printable("Failed to invalidate cache")),
+    }
```
Comment (dracarys18) [inline crates/router/src/core/cache.rs:24]: My bad i missed it but can you use `publish_and_redact_cache` function to clear the cache.

Comment (dracarys18) [reply crates/router/src/core/cache.rs:24]: Will address in a seperate PR

**Review by dracarys18 — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]
